## 2. Benchmark

In this notebook we experiment with the data by compresing it to a size reasonable for standard algorithms and then fit it to a model. This gives us a benchmark performance to compare with our Nueral Network models. 

In [6]:
#Standard imports

import numpy as np
import pandas as pd
import os
path = 'C:\\Users\\Louie\\Data\\Springboard_Cap2\\Test\\'

In [30]:
#Gather data for PCA

dfs = []

for index, file in enumerate(os.listdir(os.fsencode(path))):
    df = pd.read_csv(path + str(os.fsdecode(file)), header=None)
    dfs.append(df)
    
    if index == 9:
        break

fit = pd.concat(dfs)
fit.iloc[:,:-1].head()

0         1         2         3         4         5         6      \
0  0.278551  0.332245  0.334868  0.332245  0.338931  0.335785  0.346523   
1  0.302227  0.337168  0.337455  0.335710  0.347806  0.347931  0.347366   
2  0.343595  0.334355  0.340464  0.351578  0.343777  0.336732  0.351452   
3  0.335936  0.337815  0.341737  0.334760  0.335970  0.340356  0.338552   
4  0.341171  0.355988  0.352418  0.343288  0.345669  0.344875  0.355469   

      7         8         9        ...        13234     13235     13236  \
0  0.343990  0.346027  0.354965    ...     0.827359  0.825127  0.832944   
1  0.347366  0.355590  0.353527    ...     0.831211  0.827076  0.825912   
2  0.348867  0.354343  0.346472    ...     0.825815  0.829048  0.832970   
3  0.342499  0.352121  0.348199    ...     0.828483  0.828483  0.833253   
4  0.355469  0.355469  0.359391    ...     0.828712  0.848472  0.844169   

      13237     13238     13239     13240     13241     13242     13243  
0  0.842221  0.839381  0.848091  0.848579  0.848656  0.845545  0.851650  
1  0.836891  0.844074  0.848091  0.850331  0.850331  0.846409  0.851711  
2  0.844735  0.848091  0.851884  0.852012  0.852578  0.841624  0.855067  
3  0.836326  0.848091  0.850459  0.855556  0.848091  0.852012  0.854890  
4  0.848091  0.852578  0.854946  0.858989  0.852012  0.852012  0.856499  

[5 rows x 13244 columns]

In [31]:
#prime PCA model for transforms

from sklearn.decomposition import PCA
pca = PCA(n_components=10)
pca.fit(fit.iloc[:,:-1])

#Save precious RAM!!
fit = None 
dfs = None
df = None

In [36]:
#Transform all data to 10 dimensions using PCA

dfs = []
series = []

for index, file in enumerate(os.listdir(os.fsencode(path))):
    df = pd.read_csv(path + str(os.fsdecode(file)), header=None)
    series.append(df[13244])
    dfs.append(pd.DataFrame(pca.transform(df.iloc[:,:-1])))

data = pd.concat(dfs)
labels = pd.concat(series)

In [37]:
data.shape

(650, 10)

In [38]:
labels.shape

(650,)

In [44]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data, labels.astype(int), test_size=0.33)

In [47]:
#Fit and Score our model!

from sklearn.ensemble import RandomForestClassifier

rfc = RandomForestClassifier(n_estimators=100, max_depth=2)
rfc.fit(X_train, y_train)
score = rfc.score(X_test, y_test)

print(score)

0.8744186046511628


### Model Performance?

As we can see the model has a somewhat remarkable 87.4% accuracy! We could tweak this pipeline by adjusting the dimensionality of the data, tuning different models, and even using all the available data instead of the small sample of 650 images. This may increase performance but here we are using this as a way to compare the effectiveness of our Nueral Network models on this complex problem.